pkg/cloudproduct/gke/gke.go

```diff
@@ -84,3 +90,74 @@ func (*gkeAutopilot) SyncPodPortsToGameServer(gs *agonesv1.GameServer, pod *core
 	}
 	return nil
 }
+
+func (*gkeAutopilot) NewPortAllocator(minPort, maxPort int32,
+	_ informers.SharedInformerFactory,
+	_ externalversions.SharedInformerFactory) portallocator.Interface {
+	return &autopilotPortAllocator{minPort: minPort, maxPort: maxPort}
+}
+
+func (*gkeAutopilot) ValidateGameServer(gs *agonesv1.GameServer) []metav1.StatusCause {
+	var causes []metav1.StatusCause
+	for _, p := range gs.Spec.Ports {
+		if p.PortPolicy != agonesv1.Dynamic {
+			causes = append(causes, metav1.StatusCause{
+				Type:    metav1.CauseTypeFieldValueInvalid,
+				Field:   fmt.Sprintf("%s.portPolicy", p.Name),
+				Message: errPortPolicyMustBeDynamic,
+			})
+		}
+	}
+	return causes
+}
+
+type autopilotPortAllocator struct {
+	minPort int32
+	maxPort int32
+}
+
+func (*autopilotPortAllocator) Run(_ context.Context) error        { return nil }
+func (*autopilotPortAllocator) DeAllocate(gs *agonesv1.GameServer) {}
+
+func (apa *autopilotPortAllocator) Allocate(gs *agonesv1.GameServer) *agonesv1.GameServer {
+	if len(gs.Spec.Ports) == 0 {
+		return gs // Nothing to do.
+	}
+
+	var ports []agonesv1.GameServerPort
+	for i, p := range gs.Spec.Ports {
+		if p.PortPolicy != agonesv1.Dynamic {
+			logger.Errorf("GameServer %s port %v has PortPolicy %q - this should have been rejected by webhooks, refusing to assign ports",
+				gs.Name, p, p.PortPolicy)
+			return gs
+		}
+		p.HostPort = int32(i + 1) // Autopilot expects _some_ host port - use a value unique to this GameServer Port.
+
+		if p.Protocol == agonesv1.ProtocolTCPUDP {
+			tcp := p
+			tcp.Name = p.Name + "-tcp"
+			tcp.Protocol = corev1.ProtocolTCP
+			ports = append(ports, tcp)
+
+			p.Name += "-udp"
+			p.Protocol = corev1.ProtocolUDP
+		}
+		ports = append(ports, p)
+	}
+
+	hpa := hostPortAssignment{Min: apa.minPort, Max: apa.maxPort}
+	hpaJSON, err := json.Marshal(hpa)
+	if err != nil {
+		logger.Errorf("Internal error marshalling hostPortAssignment %v for GameServer %s: %v", hpa, gs.Name, err)
```

Comment:
Sorry, missed this logging statement on first pass.

Nit, use structured logging:

```suggestion
		logger.WithError(err).WithField("hostPort", hpa).WithField("gs", gs.Name).Error("Internal error marshalling hostPortAssignment for GameServer")
```